### PR TITLE
[SPARK-44285] MSK IAM Support

### PR DESCRIPTION
**Pull request by nihalpot**
### What changes were proposed in this pull request?

adding a dependency for [MSK IAM ](https://github.com/aws/aws-msk-iam-auth) into the repository. While Kafka can already be used to connect to MSK, this PR introduces IAM support. This should solve customer pain points as in https://github.com/aws/aws-msk-iam-auth/issues/115. 

### Why are the changes needed?

While Kafka can already be used to connect to MSK, this PR introduces IAM support. This should solve customer pain points as in https://github.com/aws/aws-msk-iam-auth/issues/115. 

### Does this PR introduce _any_ user-facing change?

Yes, customers can now use the configs like 
```
--conf spark.kafka.clusters.${cluster}.kafka.security.protocol=SASL_SSL
    --conf spark.kafka.clusters.${cluster}.kafka.sasl.mechanism=AWS_MSK_IAM
    --conf spark.kafka.clusters.${cluster}.kafka.sasl.jaas.config=software.amazon.msk.auth.iam.IAMLoginModule required;
    --conf spark.kafka.clusters.${cluster}.kafka.sasl.client.callback.handler.class=software.amazon.msk.auth.iam.IAMClientCallbackHandler
```

### How was this patch tested?

Testing is challenging since there's no easy to mock AWS locally for testing IAM. Instead we test that a specific error is thrown in various scenarios to validate that the library is discoverable. This has also been manually tested.

The stackPR for the timeout is as follows:

This is for source-side errors;
```
[info] - test MSK IAM auth on kafka 'source and stream' side *** FAILED *** (14 seconds, 971 milliseconds)
[info]   org.apache.spark.sql.streaming.StreamingQueryException: [STREAM_FAILED] Query [id = 536052c3-968d-49d4-8196-34606826a6e6, runId = 8f342aff-62bb-49ac-a8c7-d81dc4b34252] terminated with exception: org.apache.kafka.common.errors.TimeoutException: Timed out waiting for a node assignment. Call: describeTopics
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution.org$apache$spark$sql$execution$streaming$StreamExecution$$runStream(StreamExecution.scala:327)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution$$anon$1.run(StreamExecution.scala:207)
[info]   Cause: java.util.concurrent.ExecutionException: org.apache.kafka.common.errors.TimeoutException: Timed out waiting for a node assignment. Call: describeTopics
[info]   at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
[info]   at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1908)
[info]   at org.apache.kafka.common.internals.KafkaFutureImpl.get(KafkaFutureImpl.java:165)
[info]   at org.apache.spark.sql.kafka010.ConsumerStrategy.retrieveAllPartitions(ConsumerStrategy.scala:66)
[info]   at org.apache.spark.sql.kafka010.ConsumerStrategy.retrieveAllPartitions$(ConsumerStrategy.scala:65)
[info]   at org.apache.spark.sql.kafka010.SubscribeStrategy.retrieveAllPartitions(ConsumerStrategy.scala:102)
[info]   at org.apache.spark.sql.kafka010.SubscribeStrategy.assignedTopicPartitions(ConsumerStrategy.scala:113)
[info]   at org.apache.spark.sql.kafka010.KafkaOffsetReaderAdmin.$anonfun$partitionsAssignedToAdmin$1(KafkaOffsetReaderAdmin.scala:499)
[info]   at org.apache.spark.sql.kafka010.KafkaOffsetReaderAdmin.withRetries(KafkaOffsetReaderAdmin.scala:518)
[info]   at org.apache.spark.sql.kafka010.KafkaOffsetReaderAdmin.partitionsAssignedToAdmin(KafkaOffsetReaderAdmin.scala:498)
[info]   at org.apache.spark.sql.kafka010.KafkaOffsetReaderAdmin.fetchEarliestOffsets(KafkaOffsetReaderAdmin.scala:288)
[info]   at org.apache.spark.sql.kafka010.KafkaMicroBatchStream.$anonfun$getOrCreateInitialPartitionOffsets$1(KafkaMicroBatchStream.scala:244)
[info]   at scala.Option.getOrElse(Option.scala:189)
[info]   at org.apache.spark.sql.kafka010.KafkaMicroBatchStream.getOrCreateInitialPartitionOffsets(KafkaMicroBatchStream.scala:241)
[info]   at org.apache.spark.sql.kafka010.KafkaMicroBatchStream.initialOffset(KafkaMicroBatchStream.scala:98)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$getStartOffset$2(MicroBatchExecution.scala:457)
[info]   at scala.Option.getOrElse(Option.scala:189)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.getStartOffset(MicroBatchExecution.scala:457)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$constructNextBatch$4(MicroBatchExecution.scala:491)
[info]   at org.apache.spark.sql.execution.streaming.ProgressReporter.reportTimeTaken(ProgressReporter.scala:405)
[info]   at org.apache.spark.sql.execution.streaming.ProgressReporter.reportTimeTaken$(ProgressReporter.scala:403)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution.reportTimeTaken(StreamExecution.scala:67)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$constructNextBatch$2(MicroBatchExecution.scala:490)
[info]   at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
[info]   at scala.collection.Iterator.foreach(Iterator.scala:943)
[info]   at scala.collection.Iterator.foreach$(Iterator.scala:943)
[info]   at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
[info]   at scala.collection.IterableLike.foreach(IterableLike.scala:74)
[info]   at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
[info]   at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
[info]   at scala.collection.TraversableLike.map(TraversableLike.scala:286)
[info]   at scala.collection.TraversableLike.map$(TraversableLike.scala:279)
[info]   at scala.collection.AbstractTraversable.map(Traversable.scala:108)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$constructNextBatch$1(MicroBatchExecution.scala:479)
[info]   at scala.runtime.java8.JFunction0$mcZ$sp.apply(JFunction0$mcZ$sp.java:23)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.withProgressLocked(MicroBatchExecution.scala:810)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.constructNextBatch(MicroBatchExecution.scala:475)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$runActivatedStream$2(MicroBatchExecution.scala:268)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.apache.spark.sql.execution.streaming.ProgressReporter.reportTimeTaken(ProgressReporter.scala:405)
[info]   at org.apache.spark.sql.execution.streaming.ProgressReporter.reportTimeTaken$(ProgressReporter.scala:403)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution.reportTimeTaken(StreamExecution.scala:67)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$runActivatedStream$1(MicroBatchExecution.scala:249)
[info]   at org.apache.spark.sql.execution.streaming.ProcessingTimeExecutor.execute(TriggerExecutor.scala:67)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.runActivatedStream(MicroBatchExecution.scala:239)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution.$anonfun$runStream$1(StreamExecution.scala:306)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:857)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution.org$apache$spark$sql$execution$streaming$StreamExecution$$runStream(StreamExecution.scala:284)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution$$anon$1.run(StreamExecution.scala:207)
[info]   Cause: org.apache.kafka.common.errors.TimeoutException: Timed out waiting for a node assignment. Call: describeTopics
```

This is for sink-side errors (the topic is present in the metadata but it is not found in the metadata):

```
[info] Driver stacktrace:
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution.org$apache$spark$sql$execution$streaming$StreamExecution$$runStream(StreamExecution.scala:327)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution$$anon$1.run(StreamExecution.scala:207)
[info]   Cause: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 0.0 failed 1 times, most recent failure: Lost task 0.0 in stage 0.0 (TID 0) (ip-10-110-24-27.us-west-2.compute.internal executor driver): org.apache.kafka.common.errors.TimeoutException: Topic msk-123 not present in metadata after 3000 ms.
[info]
[info] Driver stacktrace:
[info]   at org.apache.spark.scheduler.DAGScheduler.failJobAndIndependentStages(DAGScheduler.scala:2844)
[info]   at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2(DAGScheduler.scala:2780)
[info]   at org.apache.spark.scheduler.DAGScheduler.$anonfun$abortStage$2$adapted(DAGScheduler.scala:2779)
[info]   at scala.collection.mutable.ResizableArray.foreach(ResizableArray.scala:62)
[info]   at scala.collection.mutable.ResizableArray.foreach$(ResizableArray.scala:55)
[info]   at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:49)
[info]   at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:2779)
[info]   at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1(DAGScheduler.scala:1242)
[info]   at org.apache.spark.scheduler.DAGScheduler.$anonfun$handleTaskSetFailed$1$adapted(DAGScheduler.scala:1242)
[info]   at scala.Option.foreach(Option.scala:407)
[info]   at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:1242)
[info]   at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:3048)
[info]   at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2982)
[info]   at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:2971)
[info]   at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:49)
[info]   at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:984)
[info]   at org.apache.spark.SparkContext.runJob(SparkContext.scala:2340)
[info]   at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.writeWithV2(WriteToDataSourceV2Exec.scala:385)
[info]   at org.apache.spark.sql.execution.datasources.v2.V2TableWriteExec.writeWithV2$(WriteToDataSourceV2Exec.scala:359)
[info]   at org.apache.spark.sql.execution.datasources.v2.WriteToDataSourceV2Exec.writeWithV2(WriteToDataSourceV2Exec.scala:307)
[info]   at org.apache.spark.sql.execution.datasources.v2.WriteToDataSourceV2Exec.run(WriteToDataSourceV2Exec.scala:318)
[info]   at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result$lzycompute(V2CommandExec.scala:43)
[info]   at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.result(V2CommandExec.scala:43)
[info]   at org.apache.spark.sql.execution.datasources.v2.V2CommandExec.executeCollect(V2CommandExec.scala:49)
[info]   at org.apache.spark.sql.Dataset.collectFromPlan(Dataset.scala:4384)
[info]   at org.apache.spark.sql.Dataset.$anonfun$collect$1(Dataset.scala:3625)
[info]   at org.apache.spark.sql.Dataset.$anonfun$withAction$2(Dataset.scala:4374)
[info]   at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:529)
[info]   at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:4372)
[info]   at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:118)
[info]   at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:195)
[info]   at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:103)
[info]   at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:857)
[info]   at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:65)
[info]   at org.apache.spark.sql.Dataset.withAction(Dataset.scala:4372)
[info]   at org.apache.spark.sql.Dataset.collect(Dataset.scala:3625)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$runBatch$17(MicroBatchExecution.scala:741)
[info]   at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$6(SQLExecution.scala:118)
[info]   at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:195)
[info]   at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withNewExecutionId$1(SQLExecution.scala:103)
[info]   at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:857)
[info]   at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:65)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$runBatch$16(MicroBatchExecution.scala:729)
[info]   at org.apache.spark.sql.execution.streaming.ProgressReporter.reportTimeTaken(ProgressReporter.scala:405)
[info]   at org.apache.spark.sql.execution.streaming.ProgressReporter.reportTimeTaken$(ProgressReporter.scala:403)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution.reportTimeTaken(StreamExecution.scala:67)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.runBatch(MicroBatchExecution.scala:729)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$runActivatedStream$2(MicroBatchExecution.scala:286)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.apache.spark.sql.execution.streaming.ProgressReporter.reportTimeTaken(ProgressReporter.scala:405)
[info]   at org.apache.spark.sql.execution.streaming.ProgressReporter.reportTimeTaken$(ProgressReporter.scala:403)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution.reportTimeTaken(StreamExecution.scala:67)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.$anonfun$runActivatedStream$1(MicroBatchExecution.scala:249)
[info]   at org.apache.spark.sql.execution.streaming.ProcessingTimeExecutor.execute(TriggerExecutor.scala:67)
[info]   at org.apache.spark.sql.execution.streaming.MicroBatchExecution.runActivatedStream(MicroBatchExecution.scala:239)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution.$anonfun$runStream$1(StreamExecution.scala:306)
[info]   at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
[info]   at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:857)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution.org$apache$spark$sql$execution$streaming$StreamExecution$$runStream(StreamExecution.scala:284)
[info]   at org.apache.spark.sql.execution.streaming.StreamExecution$$anon$1.run(StreamExecution.scala:207)
[info]   Cause: org.apache.kafka.common.errors.TimeoutException: Topic msk-123 not present in metadata after 3000 ms.
```

### CODE REVIEW

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala [hunk 1]
```diff
@@ -18,30 +18,108 @@
...
+    }
+  }
+
   test("batch mode - options should be handled as case-insensitive") {
     verifyFieldsInBatch(KafkaSourceProvider.CONSUMER_POLL_TIMEOUT, expected, batch => {
       assert(expected.toLong === batch.pollTimeoutMs)
     })
   }
 
+  /* if testType contains source/sink, kafka is used as a source/sink option respectively
```
Comment (rangadi) [inline connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:68]: What is this about? 

Could you add a brief description of the test here?

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala [hunk 2]
```diff
@@ -18,30 +18,108 @@
...
 import org.apache.spark.sql.connector.read.Scan
+import org.apache.spark.sql.functions.col
+import org.apache.spark.sql.streaming.StreamingQueryException
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.{StringType, StructType}
 import org.apache.spark.sql.util.CaseInsensitiveStringMap
 
-class KafkaSourceProviderSuite extends SparkFunSuite {
+class KafkaSourceProviderSuite extends SparkFunSuite with SharedSparkSession {
+  import scala.collection.JavaConverters._
```
Comment (rangadi) [inline connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:35]: Why is this moved inside?

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala [hunk 3]
```diff
@@ -62,6 +140,7 @@ class KafkaSourceProviderSuite extends SparkFunSuite {
         stream => {
       assert(expected.toLong === stream.kafkaOffsetReader.offsetFetchAttemptIntervalMs)
     })
+    SparkEnv.set(null)
```
Comment (anishshri-db) [inline connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:143]: Is this to fix some test failure ?

File: docs/structured-streaming-kafka-integration.md [hunk 4]
```diff
@@ -1171,3 +1171,14 @@ This can be done several ways. One possibility is to provide additional JVM para
...
+
+### IAM authentication
+
+Managed Streaming for Kafka (MSK) is a fully managed service offered by Amazon Web Services (AWS) that simplifies the deployment, management, and scaling of Apache Kafka clusters. 
+One way to connect to MSK is by using IAM authentication. To facilitate this, Spark already ships with the [aws msk library](https://github.com/aws/aws-msk-iam-auth) by default so no additional dependencies are required.
+To use IAM authentication, you can set the kafka config as in the [library documentation](https://github.com/aws/aws-msk-iam-auth). For example:
+
+    --conf spark.kafka.clusters.${cluster}.kafka.security.protocol=SASL_SSL
+    --conf spark.kafka.clusters.${cluster}.kafka.sasl.mechanism=AWS_MSK_IAM
+    --conf spark.kafka.clusters.${cluster}.kafka.sasl.jaas.config=software.amazon.msk.auth.iam.IAMLoginModule required;
```
Comment (anishshri-db) [inline docs/structured-streaming-kafka-integration.md:1183]: Is the `required;` at the end intentional ?

File: docs/structured-streaming-kafka-integration.md [hunk 5]
```diff
@@ -1171,3 +1171,14 @@ This can be done several ways. One possibility is to provide additional JVM para
...
         --conf spark.executor.extraJavaOptions=-Djava.security.auth.login.config=/path/to/custom_jaas.conf \
         ...
+
+### IAM authentication
+
+Managed Streaming for Kafka (MSK) is a fully managed service offered by Amazon Web Services (AWS) that simplifies the deployment, management, and scaling of Apache Kafka clusters. 
+One way to connect to MSK is by using IAM authentication. To facilitate this, Spark already ships with the [aws msk library](https://github.com/aws/aws-msk-iam-auth) by default so no additional dependencies are required.
+To use IAM authentication, you can set the kafka config as in the [library documentation](https://github.com/aws/aws-msk-iam-auth). For example:
+
+    --conf spark.kafka.clusters.${cluster}.kafka.security.protocol=SASL_SSL
```
[the remark this answers is not in the document]
Comment (anishshri-db) [reply docs/structured-streaming-kafka-integration.md:1181]: +1 - can we document how these options can be used in a streaming query ?

```
options = {
    "kafka.sasl.mechanism": "AWS_MSK_IAM",
    "kafka.sasl.jaas.config = "kafka.sasl.jaas.config=software.amazon.msk.auth.iam.IAMLoginModule",
    "kafka.sasl.client.callback.handler.class"="software.amazon.msk.auth.iam.IAMClientCallbackHandler",
    "kafka.security.protocol" : "SASL_SSL",
    "kafka.bootstrap.servers": bootstrap_servers,
    "group.id": group_id,
    "subscribe": topic,
}
df = spark.readStream.format("kafka").options(**options).load()
```

Maybe something like the above works ?

[the remark this answers is not in the document]
Comment (anishshri-db) [reply docs/structured-streaming-kafka-integration.md:1181]: I believe the method described below applies to `spark-submit` ?

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala [hunk 6]
```diff
@@ -18,30 +18,108 @@
...
+      )
+
+      var e: Throwable = null
+      if (testType.contains("stream")) {
+        if (testType.contains("source")) {
+          e = intercept[StreamingQueryException] {
+            spark.readStream.format("kafka").options(options).load()
+              .writeStream.format("console").start().processAllAvailable()
+          }
+          TestUtils.assertExceptionMsg(e, "Timed out")
```
Comment (anishshri-db) [inline connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:92]: Can you add a comment here ? We expect a time out because the given servers/broker don't exist ? How does it exercise the IAM path ?

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala [hunk 7]
```diff
@@ -18,30 +18,108 @@
...
+          e = intercept[StreamingQueryException] {
+            spark.readStream.format("rate").option("rowsPerSecond", 10).load()
+              .withColumn("value", col("value").cast(StringType)).writeStream
+              .format("kafka").options(options).option("checkpointLocation", "temp/testing")
+              .option("topic", "msk-123").start().processAllAvailable()
+          }
+          TestUtils.assertExceptionMsg(e, "Timeout")
+        }
+      } else {
+        if (testType.contains("source")) {
```
Comment (anishshri-db) [inline connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:103]: I would prefer to have a separate batch and streaming test, unless there is precedent in this file that does otherwise ?

File: docs/structured-streaming-kafka-integration.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (nihalpot) [reply docs/structured-streaming-kafka-integration.md:1181]: Yeah I believe that's how we set spark-confs for `spark-submit`. To make it easier to understand, I have added the syntax with `spark.readStream(...)`

File: docs/structured-streaming-kafka-integration.md (hunk 4 above)
Comment (nihalpot) [reply docs/structured-streaming-kafka-integration.md:1183]: Yup this syntax is required by the library

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala (hunk 3 above)
Comment (nihalpot) [reply connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:143]: Yeah this was initially part of the `beforeAll` command on line 35 but I removed it since the new tests rely on the spark env config set by the `SharedSparkSession`, I moved it here since the spark config gets modified by the methods in this test. Alternatively, we could set it back to what it was at the start which seems like a safer option.

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala (hunk 7 above)
Comment (nihalpot) [reply connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:103]: I think that's a fair point, there's a `DontFailOnDataLoss` suite in the same folder which incorporates both a streaming and batch test so do you think it's okay to leave it as is? Having said that, my initial motivation with this structure was to avoid code duplication and have the MSK tests in one area

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala (hunk 6 above)
Comment (nihalpot) [reply connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:101]: Added a comment in the test description. Although the broker does exist but it doesn't have IAM capabilities so the connection times out and throws an error. We are testing in this case that these specific classes in the IAM path are discoverable, which is how users interact with it (by specifying the class paths).

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala [hunk 8]
```diff
@@ -18,30 +18,124 @@
...
+      testUtils.createTopic(options("subscribe"))
+
+      var e: Throwable = null
+      if (testType.contains("stream")) {
+        if (testType.contains("source")) {
+          e = intercept[StreamingQueryException] {
+            spark.readStream.format("kafka").options(options).load()
+              .writeStream.format("console").start().processAllAvailable()
+          }
+          TestUtils.assertExceptionMsg(e, "Timed out waiting for a node assignment")
```
Comment (nihalpot) [inline connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:104]: To clarify, this timeout message is indeed expected. Essentially, the kafka client keeps calling the 'describeTopics' endpoint on the broker while using IAM authentication which times out since it doesn't have IAM auth enabled. For instance, if we remove the `jaasConfig` and `handler.class` options, this will work.

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala [hunk 9]
```diff
@@ -18,30 +18,124 @@
...
+        } else {
+          e = intercept[StreamingQueryException] {
+            spark.readStream.format("rate").option("rowsPerSecond", 10).load()
+              .withColumn("value", col("value").cast(StringType)).writeStream
+              .format("kafka").options(options).option("checkpointLocation", "temp/testing")
+              .option("topic", options("subscribe")).start().processAllAvailable()
+          }
+          TestUtils.assertExceptionMsg(e, s"TimeoutException: Topic ${options("subscribe")} " +
```
Comment (nihalpot) [inline connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:112]: ^ Read above for clarification, this is the same error, even though the topic exists, it doesn't find it in the metadata since the node doesn't offer that mode of authentication

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala (hunk 7 above)
Comment (WweiL) [reply connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:119]: I agree with Anish, you could create a helper function like `verifyMissingOffsetsDontCauseDuplicatedRecords` in the suite you mentioned: https://github.com/apache/spark/blob/fcc846e82bfb5cea61361b973ec382a5bd07c87a/connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaDontFailOnDataLossSuite.scala#L97

and either create 4 test cases like these https://github.com/apache/spark/blob/fcc846e82bfb5cea61361b973ec382a5bd07c87a/connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaDontFailOnDataLossSuite.scala#L136-L160

or use the `Seq.foreach` way. Just to align the code style

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala (hunk 8 above)
Comment (WweiL) [reply connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:107]: > Essentially, the kafka client keeps calling the 'describeTopics' endpoint on the broker while using IAM authentication which times out since it doesn't have IAM auth enabled.

If we want to keep the timeout error here I think it's good to add this "keeps calling and time out" explanation also to the comment above

Comment (WweiL) [reply connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:107]: I think it's fine at this moment since we just want to verify it works or not. In real implementation not sure if we could catch that error somewhere and throw a nicer one

**Review by WweiL — COMMENTED** [review]
Thanks for the change! Left a few comments

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala [hunk 10]
```diff
@@ -18,30 +18,127 @@
...
   override protected def afterEach(): Unit = {
-    SparkEnv.set(null)
     super.afterEach()
   }
 
+  override protected def beforeAll(): Unit = {
+    super.beforeAll()
+    testUtils = new KafkaTestUtils
+    testUtils.setup()
+  }
```
[the remark this answers is not in the document]
Comment (nihalpot) [reply connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:49]: We could that do that but I think these tests better belong to the `KafkaSourceProviderSuite` than the `KafkaMicroBatchSourceSuite` since that test suite contains more of the test suites which are testing the different options and this is primarily introducing new options which we intend to test.

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala [hunk 11]
```diff
@@ -18,30 +18,127 @@
...
+        "kafka.security.protocol" -> "SASL_SSL",
+        "kafka.sasl.client.callback.handler.class" ->
+          "software.amazon.msk.auth.iam.IAMClientCallbackHandler",
+        "retries" -> "0",
+        "kafka.request.timeout.ms" -> "3000",
+        "kafka.default.api.timeout.ms" -> "3000",
+        "kafka.max.block.ms" -> "3000"
+      )
+
+      testUtils.createTopic(options("subscribe"))
```
[the remark this answers is not in the document]
Comment (nihalpot) [reply connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:98]: We could do this but I am inclined to think it would be better to avoid this if we are just introducing this for this these tests, the other tests in this file just use a "dummy" topic. 

What do you think?

**Comment by rangadi** [comment]
Can you remove 'WIP' in the title?

**Comment by nihalpot** [comment]
> Can you remove 'WIP' in the title?

Might be an older version of the PR I think

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala [hunk 12]
```diff
@@ -66,77 +66,78 @@ class KafkaSourceProviderSuite extends SparkFunSuite with SharedSparkSession {
...
+    testUtils.deleteTopic(mskIAMTestKafkaOptions("subscribe"))
+  }
+
+  test("test MSK IAM auth with streaming API and with kafka sink") {
+    testUtils.createTopic(mskIAMTestKafkaOptions("subscribe"))
+    val e = intercept[StreamingQueryException] {
+      spark.readStream.format("rate").option("rowsPerSecond", 10).load()
+        .withColumn("value", col("value").cast(StringType)).writeStream
+        .format("kafka").options(mskIAMTestKafkaOptions).option("kafka.bootstrap.servers",
+        testUtils.brokerAddress).option("checkpointLocation", "temp/testing")
```
Comment (WweiL) [inline connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:109]: better to use a `withTempDir` method even if this location won't be created

File: connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (WweiL) [reply connector/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaSourceProviderSuite.scala:49]: Yea that makes sense

**Comment by gcoffmanQ** [comment]
@nihalpot Is there a reason that this PR was closed?

**Comment by rangadi** [comment]
@gcoffmanQ this is merged into runtime. https://docs.databricks.com/en/structured-streaming/kafka.html#use-amazon-managed-streaming-for-kafka

**Comment by ignitz** [comment]
And for Spark Vanilla (open source)?
